ArduCopter/control_safe_rtl.cpp

```diff
@@ -0,0 +1,136 @@
+ * Once the copter is close to home, it will run a standard land controller.
+ */
+
+bool Copter::safe_rtl_init(bool ignore_checks)
+{
+    Vector3f current_pos;
+    if ((position_ok() || ignore_checks) && g2.safe_rtl.is_active() && ahrs.get_relative_position_NED_origin(current_pos)) {
+        // initialise waypoint and spline controller
+        wp_nav->wp_and_spline_init();
+
+        // if we're within 4m of the landing point, just skip right to pre-land-positioning
+        // otherwise, begin a normal safertl procedure from the start
+        if (HYPOT(current_pos, g2.safe_rtl.get_point(0)) <= 4.0f) {
+            current_pos[2] -= 2.0f; // go to the point two meters above the landing spot, then start landing.
+            safe_rtl_state = SafeRTL_PreLandPosition;
+        } else {
+            safe_rtl_state = SafeRTL_WaitForPathCleanup;
+        }
+
+        // set current target to a reasonable stopping point
+        Vector3f stopping_point;
+        pos_control->get_stopping_point_xy(stopping_point);
+        pos_control->get_stopping_point_z(stopping_point);
+        wp_nav->set_wp_destination(stopping_point);
+
+        // initialise yaw to obey user parameter
+        set_auto_yaw_mode(get_default_auto_yaw_mode(true));
+
+        return true;
+    } else {
+        return false;
+    }
+}
+
+void Copter::safe_rtl_run()
+{
+    switch(safe_rtl_state){
+        case SafeRTL_WaitForPathCleanup:
+            safe_rtl_wait_cleanup_run();
+            break;
+        case SafeRTL_PathFollow:
+            safe_rtl_path_follow_run();
+            break;
+        case SafeRTL_PreLandPosition:
+            safe_rtl_pre_land_position_run();
+            break;
+        case SafeRTL_Descend:
+            rtl_descent_run(); // Re-using the descend method from normal rtl mode.
+            break;
+        case SafeRTL_Land:
+            rtl_land_run(); // Re-using the land method from normal rtl mode.
+            break;
+    }
+}
+
+void Copter::safe_rtl_wait_cleanup_run()
+{
+    // hover at current target position
+    motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
+    wp_nav->update_wpnav();
+    pos_control->update_z_controller();
+    attitude_control->input_euler_angle_roll_pitch_yaw(wp_nav->get_roll(), wp_nav->get_pitch(), get_auto_heading(),true, get_smoothing_gain());
+
+    // check if return path is computed and if yes, begin journey home
+    if (g2.safe_rtl.thorough_cleanup()) {
+        safe_rtl_state = SafeRTL_PathFollow;
+    }
+}
+
+void Copter::safe_rtl_path_follow_run()
+{
+    // if we are close to current target point, switch the next point to be our target.
+    if (wp_nav->reached_wp_destination()) {
+        Vector3f next_point;
+        bool last_point = g2.safe_rtl.pop_point(next_point);
+        if (!last_point) {
+            // go to next point along path
+            wp_nav->set_wp_destination_NED(next_point);
+        } else {   // if this is the last point, we should prepare to land
```

Comment:
i'm a little torn on what to do about the last point.  It's likely to be on the ground so flying straight towards could easily mean we hit the ground before reaching it and then flip over.